### PR TITLE
AAC-325 Defendant-Provider relationship on Hearing days

### PR DESCRIPTION
**Pull request by Obsiye**
#### What

Add the Provider(Defence counsel) to Defendant relationship on Hearing days pages

#### Ticket

[AAC-325](https://dsdmoj.atlassian.net/browse/AAC-325)

#### Why

In order to create the provider to defendant relationship on the hearing day pages.

#### How

- Map the defendant_ids in the hearing to the defendant details in the prosecution cases
  - hearing > [defence_counsel] > [defendants]
  - hearing > [prosecution cases] > [defendants] > defendant details > person details
⚠️ the prosecution cases haven't been filtered to the case being used when arriving on the hearing day page. However, I assume the defendant details/defendant ids are the same across hearing.prosecution_cases
- When there are defendant ids but can't match to defendant details it becomes John Smith (QC) for not available

- Add a DefenceCounselListService to build the provider/defendant relationship
- Duplicate the defence counsel list when the defence counsel has multiple defendants:
  - e.g. John Smith (QC) for Defendant1< br>John Smith (QC) for Defendant2
- Guard against if a defendant id fails to match a defendant on both the hearing page and prosecution case page (defence counsels) e.g. John Smith (QC) for not available
- Capitalise the first name, middle name and last name of Defence counsels and their defendants
- Add unit tests and update integration test
--------

image

### CODE REVIEW

**Review by SamKyprianou — DISMISSED** [review]
LGTM

**Review by ivanELEC — DISMISSED** [review]

**Review by SamKyprianou — DISMISSED** [review]

File: app/decorators/cd_api/defence_counsel_decorator.rb [hunk 1]
```diff
@@ -22,20 +22,22 @@ def name_and_status_blank?
...
+      [first_name, middle_name, last_name].filter_map { |n| n&.capitalize }.reject(&:empty?).join(' ')
     end
 
     def formatted_name
       name || t('generic.not_available')
     end
 
     def formatted_status
-      status || t('generic.not_available').downcase
+      status || I18n.t('generic.not_available').downcase
```
Comment (pking-moj) [inline app/decorators/cd_api/defence_counsel_decorator.rb:33]: If we have to add `I18n` then it needs adding to the method above

File: app/decorators/cd_api/hearing_decorator.rb [hunk 2]
```diff
@@ -6,5 +6,41 @@ def cracked_ineffective_trial
       @cracked_ineffective_trial ||= decorate(object.hearing.cracked_ineffective_trial,
                                               CdApi::CrackedIneffectiveTrialDecorator)
     end
+
+    def defence_counsels_list
+      # TODO: Add logic of attendance days for each defence
```
Comment (pking-moj) [inline app/decorators/cd_api/hearing_decorator.rb:11]: Can you confirm we have a ticket associated with this TODO please

File: app/decorators/cd_api/hearing_decorator.rb [hunk 3]
```diff
@@ -6,5 +6,41 @@ def cracked_ineffective_trial
...
+      return t('generic.not_available') if hearing.defence_counsels.blank?
```
Comment (pking-moj) [inline app/decorators/cd_api/hearing_decorator.rb:12]: Do we need `I18n` here?

File: app/services/cd_api/hearing_details/defence_counsels_list_service.rb [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+          person_details = defendant.defendant_details&.person_details
+          defendant_name = format_defendant_name(**person_details&.attributes&.symbolize_keys)
+
+          @result << ("#{sentence} for #{defendant_name}")
+        end
+      end
+
+      def formatted_name(person_details)
+        name(**person_details) || t('generic.not_available')
```
Comment (pking-moj) [inline app/services/cd_api/hearing_details/defence_counsels_list_service.rb:49]: Do we need to be using `I18n` here?

File: app/decorators/cd_api/hearing_decorator.rb (hunk 2 above)
Comment (Obsiye) [reply app/decorators/cd_api/hearing_decorator.rb:11]: Yep, it's the ticket [AAC-828](https://dsdmoj.atlassian.net/browse/AAC-828?atlOrigin=eyJpIjoiZTE4MGNiZmE2NjZiNDVhNmE1MGZlMjRjMGI2MWYzNWYiLCJwIjoiaiJ9)

File: app/decorators/cd_api/defence_counsel_decorator.rb (hunk 1 above)
Comment (Obsiye) [reply app/decorators/cd_api/defence_counsel_decorator.rb:33]: Not required here, sorry. will delete

File: app/decorators/cd_api/hearing_decorator.rb (hunk 3 above)
Comment (Obsiye) [reply app/decorators/cd_api/hearing_decorator.rb:12]: not required in the decorators

File: app/services/cd_api/hearing_details/defence_counsels_list_service.rb (hunk 4 above)
Comment (Obsiye) [reply app/services/cd_api/hearing_details/defence_counsels_list_service.rb:49]: yep needed, adding now 👍🏾

**Review by pking-moj — DISMISSED** [review]

**Review by pking-moj — APPROVED** [review]

**Pull request merged**